### ISSUE TITLE
a constrained TypeVar should be assignable to another constrained TypeVar with a superset of its constraints

### ISSUE THREAD

**Issue by neutrinoceros** [astral-sh/ty#2728]
### Summary

Here's a minimal example:
`src/pkg` contains
```
__init__.py # empty
types.py
a.py
b.py
```

with

```py
# types.py
from numpy import float32 as f32
from numpy import float64 as f64
from typing import TypeVar

F = TypeVar("F", f32, f64)
```

```py
# a.py
from .types import F
from numpy.typing import NDArray

def a(a: NDArray[F]) -> NDArray[F]:
    return a
````

```py
# b.py
from .a import a
from numpy.typing import NDArray
from numpy import float32 as f32
from numpy import float64 as f64
from typing import TypeVar

B = TypeVar("B", f32, f64)

def b(b: NDArray[B]) -> NDArray[B]:
    return a(b)
```

running `ty check src` gives me
```
error[invalid-argument-type]: Argument to function `a` is incorrect
  --> src/reprod_ty/b.py:11:14
   |
10 | def b(b: NDArray[B]) -> NDArray[B]:
11 |     return a(b)
   |              ^ Argument type `B@b` does not satisfy constraints (`floating[_32Bit]`, `float64`) of type variable `F`
   |
info: Type variable defined here
 --> src/reprod_ty/types.py:5:1
  |
3 | from typing import TypeVar
4 |
5 | F = TypeVar("F", f32, f64)
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
info: rule `invalid-argument-type` is enabled by default
```

`B` and `F` are distinct, but actually equivalent type vars. I would expect `B` to behave as an adequate substitute for `F`.

This comes from a real life case where `F` and `a` are defined in some external package, with `a` being part of the public interface but `F` is *not* (type definitions are private), so I wanted to replicate the type's definition in code I control in `b.py`. FWIW, mypy doesn't reject this code, though I don't consider this proof that it *shouldn't*. Maybe there's something off in my understanding of how typevars are supposed to work ?

### Version

ty 0.0.15 (eb43dff06 2026-02-04)

**Comment by neutrinoceros** [comment]
(I don't expect numpy to actually be important here, but I have yet to try removing it)

**Comment by neutrinoceros** [comment]
A slightly simpler example (although the reported violation is different)
```py
# types.py
from typing import TypeVar

F = TypeVar("F", int, float)
```
```py
# a.py
from .types import F
from numpy.typing import NDArray

def a(a: F) -> F:
    return a
```
```py
# b.py
from .a import a
from typing import TypeVar

B = TypeVar("B", int, float)

def b(b: B) -> B:
    return a(b)
```

output:
```
error[invalid-return-type]: Return type does not match returned value
 --> src/reprod_ty/b.py:7:16
  |
5 | B = TypeVar("B", int, float)
6 |
7 | def b(b: B) -> B:
  |                - Expected `B@b` because of return type
8 |     return a(b)
  |            ^^^^ expected `B@b`, found `int | float`
  |
info: rule `invalid-return-type` is enabled by default
```

**Comment by carljm** [comment]
Thanks for the report! I think your second example has different results because `float` means `int | float`, so the one constraint is a subtype of the other. So that means we decide that `B` is a subtype of `float`, which is one of the constraints of `F`, so we solve the typevar to that constraint.

Here's a simplified repro of the original issue, that shows the original form of the error:

```py
from typing import TypeVar

T1 = TypeVar("T1", int, str)
T2 = TypeVar("T2", int, str)

def f1(x: T1) -> T1:
    return x

def f2(x: T2) -> T2:
    return f1(x)  # incorrect error
```

**Comment by carljm** [comment]
We do handle this correctly for typevars with upper bound, just not for constrained type vars:

```py
from typing import TypeVar

T1 = TypeVar("T1", bound=int)
T2 = TypeVar("T2", bound=int)

def f1(x: T1) -> T1:
    return x

def f2(x: T2) -> T2:
    return f1(x)  # works
```

**Comment by neutrinoceros** [comment]
For what it's worth I just tried ty 0.0.16 on the code from which this report originated and I'm not seeing any change wrt 0.0.15's diagnostics. I see https://github.com/astral-sh/ruff/pull/23103 is explicitly listed on the changelog though, so I'm not sure what to conclude. Is it possible that the patch was accidentally excluded from the release ? Does it mean my reproducer didn't actually capture the real life error I got ? Something else ?

**Comment by carljm** [comment]
The original three-file example using numpy that was reported in this issue passes with no errors for me using `uvx ty@0.0.16 check`, and fails with the originally reported error using `uvx ty@0.0.15 check`. So I think the issue reported here is fixed, and that fix was released in 0.0.16.

> Does it mean my reproducer didn't actually capture the real life error I got ?

Yeah, that seems like the most likely possibility.

One thing to note (called out in the tests added in https://github.com/astral-sh/ruff/pull/23103) is that we require the constraints to be equivalent types (there can be a subset of constraints, but _subtype_ constraints are not allowed). This is necessary for soundness, but some other type checkers allow it unsoundly. So it's possible that your real case involves constraints that don't match exactly, but where one is a subtype of the other?

**Comment by neutrinoceros** [comment]
thanks !

> it's possible that your real case involves constraints that don't match exactly, but where one is a subtype of the other?

unfortunately no, my real-life case actually has two typevars exactly identical in definitions and actually equal to `TypeVar("F", f32, f64)`
I'll do a couple additional checks real quick and get back to you in a couple minutes

**Comment by neutrinoceros** [comment]
I could confirm that my original example typechecks on my system with ty 0.0.16 (which essentially just proves I'm really running the version I claimed to, but I'll take it). I'm maybe just really confused and tired, so I think I better just try this again tomorrow on a clear mind, but in the mean time, just in case you're curious, the real life reprod is
```
gh repo clone la-niche/nonos
cd nonos
uv sync --all-extras
ty check src
```

**Comment by neutrinoceros** [comment]
Ok now I'm almost sure that the remaining issues I'm seeing aren't bugs in ty, rather, problems with type annotations on my side, that are looser than I thought.
I also got confused yesterday because I forgot the actual package I was working on initially was `la-niche/lick`, not `la-niche/nonos` (which has its own issues with type checking). Anyway, just reporting back that I don't anticipate finding anything else broken in ty here; but if I do, count on me to report. Thanks for your assistance and prompt fix, much appreciated !
